cache for float matching

encryptedcommerce opened the pull request:
Fixes #164 

Reduces unnecessary multiple calls to _float_matching by storing results of previous function calls in a module-level cache.

Eliminates multiple warnings during recursive use of _float_matching.

Standardizes use of `col` for column name and `X_col` for the actual column.

amueller commented:
Thanks! I've toyed with the idea of passing around a data structure that's more rich than just the dataframe, so that could also be a solution. But that would be much more invasive so I think this is a good immediate solution.

Can you maybe add a regression test that tests that the warning is only raised once?

encryptedcommerce commented:
Unfortunately I'm stuck trying to get `pytest tests/test_models.py` to pass.  I've committed my debugging output.  It seems for many cases the cache payload the the function response are equal, but then at the end the values are different and mayhem ensues.

amueller commented on dabl/preprocessing.py:
```diff
@@ -27,24 +28,65 @@ def _float_matching(col, return_safe_col=False):
+    If not present in cache, stores function call results into cache.
+    Uses dataframe object id and column name as cache key.
+    """
+    hash_key = f'{id(X)}-{col}'
```
I think ``id(X)`` is not a valid way to hash a dataframe. You can change columns without changing the object identity of the dataframe. The only way to make sure is to hash the whole object, which is quite expensive. That's what joblib and dask do afaik.
If you have a more local cache that ensures the object doesn't get mutated it might work.

amueller replied:
A simple way to test this would be to do ``from joblib import hash`` and use that instead of ``id``. In practice that might be slow, though. You could do ``hash(X[col])`` I'm not sure if that's faster or slower. Maybe there's other ways to avoid the double computation.

encryptedcommerce replied:
Thanks for the tip! `hash(X[col])` worked just fine: it was a second faster than `hash(X)` and it didn't add significant overhead.

encryptedcommerce commented:
> Can you maybe add a regression test that tests that the warning is only raised once?

I saw some peculiar behavior with the number of warnings.  I was getting one warning in the notebook but two warnings in the regression test.  I was able to hone in on it a bit by printing the warning along with the value of `return_safe_col`.  I saw that upon restarting the kernel in the notebook, I could get one error from the use of `X.apply(_float_matching)` in `_find_string_floats`, and one warning from the DirtyFloatCleaner `fit_transform`.

I added a module-wide map to track mixed-type warnings for column names, and was able to reduce the warning to one instance.  However, in the regression test I added, there were now *no* warnings for some reason.  I set the assert to look for `< 2` warnings.

amueller commented:
Hm but there should be a warning, that's strange. I might have time to look at it more tomorrow, I have to work on some other things today unfortunately.

encryptedcommerce commented:
Figured it out: it's a module-wide limit, so for all the tests in test_preprocessor.py there will only be one mixed-type warning for the same column name.  There is a call stack on a prior test that issues the warning:
`test_detect_types` --> `detect_types` --> `_find_string_floats` --> `X.apply(_float_matching)`

Therefore, by the time the `test_dirty_float_single_warning` is run, there is no warning to catch.

I could simply reverse the order so that `test_dirty_float_single_warning` is run before `test_detect_types` ; that seems to work with `assert len(w) == 1`

amueller commented:
There's a limit on the warnings? On deprecation warnings? That seems weird, I don't think we ever ran into that for sklearn so I feel like something else is going on.

I don't think we should rely on the order. How did you change the order in which the tests are run?

Also, did you check how slow ``hash`` is compare to doing the check again with a very big column?

encryptedcommerce commented:
The point of this PR was to address #164: "calling clean on mixed type dirty float gives three warnings" by creating a module-level cache for the float_matching function calls in order to reduce the number of times the warning was displayed, by reducing the amount of times the same function was invoked for the same data.  This prevents multiple warnings when you run `clean()` or `fit_transform()` -- or when you run `detect_types()` and then `clean()`, since they both invoke `_float_matching()`

Changing the order of test execution was done by simply putting `test_dirty_float_single_warning` before `test_detect_types`.  If you don't want to rely on the test order, then we can simply revert to how I had it before: `test_dirty_float_single_warning` having `assert len(w) < 2` instead of `assert len(w) == 1` so that it will check for the warning being issued "at most once" instead of "exactly once"

encryptedcommerce commented:
Regarding overhead of `hash`, I had answered in a resolved conversation so it's a bit buried:

> Thanks for the tip! hash(X[col]) worked just fine: it was a second faster than hash(X) and it didn't add significant overhead.

For more specific benchmarking, I tested against a dataframe with over 10k rows.  I saw that even though hashing the entire dataframe took a second, hashing a single column (as I've done for the cache) takes less than a millisecond -- screenshot attached:
<img width="598" alt="hash-col-time" src="https://user-images.githubusercontent.com/1234749/73844412-878a5580-47ee-11ea-894f-43607a4f9c71.png">

encryptedcommerce commented:
@amueller Any further thoughts or questions on this PR?

amueller commented:
Looks good, thanks. Sorry for the slow reply!
